src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/NodeMonitoringFailoverPluginTest.java

```diff
@@ -0,0 +1,312 @@
+import com.mysql.cj.conf.HostInfo;
+import com.mysql.cj.conf.PropertyKey;
+import com.mysql.cj.conf.PropertySet;
+import com.mysql.cj.conf.RuntimeProperty;
+import com.mysql.cj.exceptions.CJCommunicationsException;
+import com.mysql.cj.log.Log;
+import com.mysql.cj.log.NullLogger;
+import org.jboss.invocation.InvocationException;
+import org.jboss.util.NullArgumentException;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Stream;
+
+class NodeMonitoringFailoverPluginTest {
+
+  @Mock
+  private NodeMonitoringFailoverPlugin.IMonitorServiceInitializer monitorServiceInitializer;
+  @Mock
+  private PropertySet propertySet;
+  @Mock
+  private HostInfo hostInfo;
+  @Mock
+  private IFailoverPlugin mockPlugin;
+  @Mock
+  private Log logger;
+  @Mock
+  private IMonitorService monitorService;
+  @Mock
+  private Callable sqlFunc;
+  @Mock
+  private MonitorConnectionContext context;
+
+  @Mock
+  private RuntimeProperty<Boolean> nativeFailureDetectionEnabledProperty;
+  @Mock
+  private RuntimeProperty<Integer> failureDetectionTimeProperty;
+  @Mock
+  private RuntimeProperty<Integer> failureDetectionIntervalProperty;
+  @Mock
+  private RuntimeProperty<Integer> failureDetectionCountProperty;
+
+
+  private static final String NODE = "node";
+  private static final String NO_MONITOR_METHOD_NAME = "foo";
+  private static final String MONITOR_METHOD_NAME = "executeQuery";
+
+  private static final int FAILURE_DETECTION_TIME = 10;
+  private static final int FAILURE_DETECTION_INTERVAL = 100;
+  private static final int FAILURE_DETECTION_COUNT = 5;
+  private static final ExecutionException EXECUTION_EXCEPTION = new ExecutionException(
+      "exception",
+      new InvocationException("exception", new Throwable()));
+
+  private NodeMonitoringFailoverPlugin plugin;
+  private AutoCloseable closeable;
+
+  @BeforeEach
+  void init() {
+    closeable = MockitoAnnotations.openMocks(this);
+    plugin = new NodeMonitoringFailoverPlugin();
+
+    initDefaultMockReturns();
+  }
+
+  @AfterEach
+  void cleanUp() throws Exception {
+    closeable.close();
+  }
+
+  @ParameterizedTest
+  @MethodSource("generateNullArguments")
+  void test_1_initWithNullArguments(
+      final PropertySet set,
+      final HostInfo info,
+      final IFailoverPlugin failoverPlugin,
+      final Log log) {
+    Assertions.assertThrows(
+        NullArgumentException.class,
+        () -> plugin.init(set, info, failoverPlugin, log));
+  }
+
+  @Test
+  void test_2_initWithFailoverEnabled() {
+    Mockito
```

Comment:
It's better to have when/do/thenReturn static import, and not need to call Mockito.when().thenReturn() every time